### PR TITLE
21 feat dagster code location implement all fields

### PR DESCRIPTION
**Pull request by nicogelders**

### CODE REVIEW

File: internal/client/types/custom.go [hunk 1]
```diff
@@ -1,15 +1,57 @@
...
+	Image            string                 `json:"image,omitempty"`
+	CodeSource       CodeLocationCodeSource `json:"code_source"`
+	WorkingDirectory string                 `json:"working_directory,omitempty"`
+	ExecutablePath   string                 `json:"executable_path,omitempty"`
+	Attribute        string                 `json:"attribute,omitempty"`
+	Git              CodeLocationGit        `json:"git,omitempty"`
+	AgentQueue       string                 `json:"agent_queue,omitempty"`
+}
+
+func (cl CodeLocation) Validate() error {
```
Comment (QuintenBruynseraede) [inline internal/client/types/custom.go:20]: Put this logic in a ConfigValidator in the resource? https://developer.hashicorp.com/terraform/plugin/framework/resources/validate-configuration

That module provides functions like [func ExactlyOneOf(expressions ...path.Expression) resource.ConfigValidator](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework-validators@v0.12.0/resourcevalidator#ExactlyOneOf)

File: internal/provider/resources/code_location.go [hunk 2]
```diff
@@ -40,23 +45,70 @@ func (r *CodeLocationResource) Schema(ctx context.Context, req resource.SchemaRe
...
 		Attributes: map[string]schema.Attribute{
 			"name": schema.StringAttribute{
-				MarkdownDescription: "Code Location name",
+				MarkdownDescription: "Code Location name. ",
 				Required:            true,
 			},
 			"image": schema.StringAttribute{
-				MarkdownDescription: "Code Location image",
-				Required:            true,
+				MarkdownDescription: "Code Location image. Git or Image is a required field (mutually exclusive).",
```
Comment (QuintenBruynseraede) [inline internal/provider/resources/code_location.go:52]:
```suggestion
				MarkdownDescription: "Docker image URL to use. Must be specified if `git` is not defined.",
```

File: internal/provider/resources/code_location.go [hunk 3]
```diff
@@ -40,23 +45,70 @@ func (r *CodeLocationResource) Schema(ctx context.Context, req resource.SchemaRe
...
+				Required:            false,
+				Optional:            true,
 			},
 			"code_source": schema.SingleNestedAttribute{
 				Attributes: map[string]schema.Attribute{
+					"module_name": schema.StringAttribute{
+						MarkdownDescription: "Code Location code source module name. One of module_name/package_name/python_file is required (mutually exclusive).",
```
Comment (QuintenBruynseraede) [inline internal/provider/resources/code_location.go:59]:
```suggestion
						MarkdownDescription: "Name of the Python module from which to load definitions. Exactly one of `module_name`, `package_name` or `python_file` is required.",
```

File: internal/provider/resources/code_location.go [hunk 4]
```diff
@@ -40,23 +45,70 @@ func (r *CodeLocationResource) Schema(ctx context.Context, req resource.SchemaRe
...
+						Required:            false,
+						Optional:            true,
+					},
+					"package_name": schema.StringAttribute{
+						MarkdownDescription: "Code Location code source package name. One of module_name/package_name/python_file is required (mutually exclusive).",
```
Comment (QuintenBruynseraede) [inline internal/provider/resources/code_location.go:64]:
```suggestion
						MarkdownDescription: "Name of the Python package from which to load definitions. Exactly one of `module_name`, `package_name` or `python_file` is required.",
```

File: internal/provider/resources/code_location.go [hunk 5]
```diff
@@ -40,23 +45,70 @@ func (r *CodeLocationResource) Schema(ctx context.Context, req resource.SchemaRe
...
+						MarkdownDescription: "Code Location code source python file. One of module_name/package_name/python_file is required (mutually exclusive).",
+						Required:            false,
+						Optional:            true,
 					},
 				},
 				MarkdownDescription: "Code Location code source",
 				Required:            true,
 			},
+			"working_directory": schema.StringAttribute{
+				MarkdownDescription: "Code Location working directory",
```
Comment (QuintenBruynseraede) [inline internal/provider/resources/code_location.go:78]: Does this have a default? Can we put it in the description?

File: internal/provider/resources/code_location.go [hunk 6]
```diff
@@ -40,23 +45,70 @@ func (r *CodeLocationResource) Schema(ctx context.Context, req resource.SchemaRe
...
+				Required:            false,
+				Optional:            true,
+			},
+			"executable_path": schema.StringAttribute{
+				MarkdownDescription: "Code Location executable path",
```
Comment (QuintenBruynseraede) [inline internal/provider/resources/code_location.go:83]: Python? Which executable?

Comment (QuintenBruynseraede) [reply internal/provider/resources/code_location.go:83]: https://docs.dagster.io/concepts/code-locations/workspace-files#examples

Comment (QuintenBruynseraede) [reply internal/provider/resources/code_location.go:83]: Might be Dagster OSS only? Maybe we don't need to support these attributes

Comment (nicogelders) [reply internal/provider/resources/code_location.go:83]: is in the dagster cloud graphql api, so we have to support it

File: internal/provider/resources/code_location.go (hunk 5 above)
Comment (nicogelders) [reply internal/provider/resources/code_location.go:78]: Don't think so.

File: internal/provider/resources/code_location.go [hunk 7]
```diff
@@ -40,27 +47,88 @@ func (r *CodeLocationResource) Schema(ctx context.Context, req resource.SchemaRe
...
+			"agent_queue": schema.StringAttribute{
+				MarkdownDescription: "Code Location agent queue",
+				Required:            false,
+				Optional:            true,
+			},
 		},
 	}
 }
 
+func (r CodeLocationResource) ConfigValidators(ctx context.Context) []resource.ConfigValidator {
```
Comment (QuintenBruynseraede) [inline internal/provider/resources/code_location.go:118]: This one is on the object and all other functions are on the pointer

File: internal/provider/resources/code_location.go [hunk 8]
```diff
@@ -125,18 +216,39 @@ func (r *CodeLocationResource) Read(ctx context.Context, req resource.ReadReques
...
+	}
+	codeSourceAttributeValues := map[string]attr.Value{
+		"module_name":  types.StringValue(codeLocation.CodeSource.ModuleName),
+		"package_name": types.StringValue(codeLocation.CodeSource.PackageName),
+		"python_file":  types.StringValue(codeLocation.CodeSource.PythonFile),
+	}
+	codeSource, diag := types.ObjectValue(codeSourceAttributeTypes, codeSourceAttributeValues)
+	resp.Diagnostics.Append(diag...)
+
+	if resp.Diagnostics.HasError() {
```
Comment (QuintenBruynseraede) [inline internal/provider/resources/code_location.go:238]: Is returning correct? That leaves the resource half-"read"?

Comment (QuintenBruynseraede) [reply internal/provider/resources/code_location.go:238]: https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/types#ObjectValueMust

> ObjectValueMust creates a Object with a known value, converting any diagnostics into a panic at runtime

Comment (nicogelders) [reply internal/provider/resources/code_location.go:238]: rather have a graceful error than a panic tbh

**Review by QuintenBruynseraede — APPROVED** [review]

**Pull request merged**